source/Octopus.Versioning/Unsortable/UnsortableVersion.cs

```diff
@@ -0,0 +1,136 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Octopus.Versioning.Unsortable
+{
+    public class UnsortableVersion: IVersion
+    {
+        public UnsortableVersion(string release, string? metadata, string? originalString)
+        {
+            Metadata = metadata;
+            Release = release;
+            OriginalString = originalString ?? string.Empty;
+        }
+
+        public int Major => 0;
+        public int Minor => 0;
+        public int Patch => 0;
+        public int Revision => 0;
+        public bool IsPrerelease => false;
+        public IEnumerable<string> ReleaseLabels => Enumerable.Empty<string>();
+        public string? Metadata { get; }
+        public bool HasMetadata => !string.IsNullOrWhiteSpace(Metadata);
+        public string Release { get; }
+        public string OriginalString { get; }
+
+        public VersionFormat Format => VersionFormat.Unsortable;
+        
+        public int CompareTo(object obj)
+        {
+            if (!(obj is IVersion objVersion))
+                return -1;
+            
+            if (string.Compare(Release.AlphaNumericOnly(), (objVersion.Release ?? string.Empty).AlphaNumericOnly(), StringComparison.Ordinal) != 0)
+                return CompareReleaseLabels(Release.AlphaNumericOnly().Split('.', '-', '_'), (objVersion.Release ?? string.Empty).AlphaNumericOnly().Split('.', '-', '_'));
+                
+            return 0;
```

Comment:
🤔 Do we get any benefit from this? I thinking we could just return `0` every time

Reply:
Ah, it sucks that IVersion is also IComparable 🤔

Reply:
Is it possible to separate the two interfaces? So that IVersion does not directly extend IComparable.
This means that you'd have to go to all existing implementations of IVersion and add IComparable to it explicitly. This should give the benefit that you won't have to implement IComparable methods on Unsortable, which itself seems kind of misleading. 

Alternatively, have another interface like `ISortableVersion` that inherits from `IVersion` and `IComparable`. That way a consumer of ISortableVersion knows for sure that whatever it is getting is sortable and can be ordered.

Reply:
I think it is possible but the effort required might be too great, there is quite a few places around Octopus and Calamari that expect `IVersion` to sortable e.g. https://github.com/OctopusDeploy/Calamari/blob/b63a7e5ed31644b85a3ade67534eae97b806b055/source/Calamari.Shared/Integration/Certificates/FileSystem/PackageStore.cs#L77 so we would end having to update all of these places well

I think `UnsortableVersion` is probably a bad name for this since it is technically sortable just only lexicographically and not by the version numbers so maybe we just need to make it obvious how it sorts, compared to the other version types. 

@zentron Any thoughts on this?

Reply:
I agree, I don't like that we call this `UnsortableVersion` but then sort it.

The way I see it we have 2 options. 
1. Actually, separate out the interfaces as suggested by @kevjt 
1. Rename this version type to something like `LexicographicSortedVersion` and continue with this approach.

Off the top of my head, I can't think what downsides we would have with using the lexicographic versioning strategy for our commit hashes. We could get some slight performance improvement from _not_ comparing something that can't be compared, but it's probably of minimal impact in the real world, all things considered.

Before moving forward with it, I'd be interested to see a quick spike in how many places would be impacted from a separate sortable interface.

Reply:
To close this off, I did a quick spike to see how much effort would be required to split the interface and it's a decent amount, there is about 130 different usages of `IVersion` within Octopus Server and there is a few spots where the usages of `IVersion` it is always expected to be compared, an example of this would be the `PackageCache` where a package could be both `IVersion` and `ISortableVersion` so we would have to make some changes to functionality and it's hard to get insight to how that change would affect other places in code.

So we are continuing with option 2